client/peerset/src/protocol_controller.rs

```diff
@@ -0,0 +1,1902 @@
+			Ok(())
+		} else {
+			// The peer was not found in neither regular or reserved lists.
+			Err(peer_id)
+		}
+	}
+
+	/// Try dropping the peer as a reserved peer. Return `Ok(true)` if the peer was found and
+	/// disconnected, `Ok(false)` if it wasn't found, `Err(PeerId)`, if the peer found, but not in
+	/// connected state.
+	fn drop_reserved_peer(&mut self, peer_id: &PeerId) -> Result<bool, PeerId> {
+		let Some(state) = self.reserved_nodes.get_mut(peer_id) else {
+			return Ok(false)
+		};
+
+		if let PeerState::Connected(direction) = state {
+			trace!(target: LOG_TARGET, "Reserved peer {peer_id} ({direction:?}) dropped.");
+			*state = PeerState::NotConnected;
+			Ok(true)
+		} else {
+			Err(*peer_id)
+		}
+	}
+
+	/// Try dropping the peer as a regular peer. Return `true` if the peer was found and
+	/// disconnected, `false` if it wasn't found.
+	fn drop_regular_peer(&mut self, peer_id: &PeerId) -> bool {
+		let Some(direction) = self.nodes.remove(peer_id) else {
+			return false
+		};
+
+		trace!(target: LOG_TARGET, "Peer {peer_id} ({direction:?}) dropped.");
+
+		match direction {
+			Direction::Inbound => self.num_in -= 1,
+			Direction::Outbound => self.num_out -= 1,
+		}
+
+		true
+	}
+
+	/// Initiate outgoing connections trying to connect all reserved nodes and fill in all outgoing
+	/// slots.
+	fn alloc_slots(&mut self) {
+		// Try connecting to reserved nodes first, ignoring nodes with outstanding events/actions.
+		self.reserved_nodes
+			.iter_mut()
+			.filter_map(|(peer_id, state)| {
+				(!state.is_connected() && !self.peer_store.is_banned(peer_id)).then(|| {
+					*state = PeerState::Connected(Direction::Outbound);
+					peer_id
+				})
+			})
+			.cloned()
+			.collect::<Vec<_>>()
+			.into_iter()
+			.for_each(|peer_id| {
+				self.start_connection(peer_id);
+			});
+
+		// Nothing more to do if we're in reserved-only mode or don't have slots available.
+		if self.reserved_only || self.num_out >= self.max_out {
+			return
+		}
+
+		// Fill available slots.
+		let available_slots = (self.max_out - self.num_out).saturated_into();
+
+		// Ignore reserved nodes (connected above), already connected nodes, and nodes with
+		// outstanding events/actions.
+		let ignored = self
+			.reserved_nodes
+			.keys()
+			.collect::<HashSet<&PeerId>>()
+			.union(&self.nodes.keys().collect::<HashSet<&PeerId>>())
+			.cloned()
+			.collect();
+
+		let candidates = self
+			.peer_store
+			.outgoing_candidates(available_slots, ignored)
+			.into_iter()
+			.filter_map(|peer_id| {
+				(!self.reserved_nodes.contains_key(&peer_id) && !self.nodes.contains_key(&peer_id))
+					.then_some(peer_id)
+					.or_else(|| {
+						error!(
+							target: LOG_TARGET,
+							"`PeerStore` returned a node we asked to ignore: {peer_id}.",
+						);
+						debug_assert!(false, "`PeerStore` returned a node we asked to ignore.");
+						None
+					})
+			})
+			.collect::<Vec<_>>();
+
+		if candidates.len() > available_slots {
```

Comment:
is it even possible?

Reply:
This is also a precaution in case of misbehaving `PeerStore`.